### ISSUE TITLE
[Feature]: Events extraction from Neuropixels data

### ISSUE THREAD

**Issue by TheChymera** [catalystneuro/neuroconv#923]
### What would you like to see added to NeuroConv?

How can we best get support for this https://github.com/NeuralEnsemble/python-neo/pull/1383 included in neuroconv?

@CodyCBakerPhD
@manimoh

### Is your feature request related to a problem?

Converting neuropixels data

### Do you have any interest in helping implement the feature?

Yes, but I would need guidance.

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/catalystneuro/neuroconv/blob/main/.github/CODE_OF_CONDUCT.rst)
- [X] Have you ensured this bug was not already [reported](https://github.com/catalystneuro/neuroconv/issues)?

**Comment by manimoh** [comment]
I imagine changes have to be made to [spikeglxnidqinterface.py](https://github.com/catalystneuro/neuroconv/blob/main/src/neuroconv/datainterfaces/ecephys/spikeglx/spikeglxnidqinterface.py), since now the information from the event channels (digital channels) is also read by Neo

**Comment by tabedzki** [comment]
@CodyCBakerPhD  and @h-mayorquin

In `spikeglxnidqinterface.py`, is there a way to manually specify the `Extractor` value? If offered that might address the problem of not having this implemented. 

I personally would like access to the `SpikeGLXEventExtractor` as it would allow for the NeuroConv interface to access the individual bits of the nidq stream which is what I need to sync my neuroconv data to my behavioral data.

**Comment by h-mayorquin** [comment]
Where this should end up in the nwbfile? How do you envision this working?

**Comment by tabedzki** [comment]
Herberto, our signal is sent in 8 bits and it is encoded as such (see image at the bottom):

- Bit 0: timing bit that square waves with a frequency of 1 second (0.5 second at 0 and 0.5 second at 1)
- Bit 1: sends a square wave when our trial starts for our mouse training session
- Bit 2: Sends a pulse for every frame that our virtual reality machine generates/sends
- Bit 3-7: Encodes a 5 bit word that serves as a counter for the simulation frame number (modulo 2^5)

> Recently the spikeglx events were implemented at the NEO level: https://github.com/NeuralEnsemble/python-neo/pull/1383 and are also available at the spikeinterface level with the read_spikeglx_events  function.

This feature does the unpacking needed and does it within the SpikeInterface code. As of right now, I would have to implement the parsing of bits in a custom aligning event through [NWBConverter.temporally_align_data_interfaces()](https://neuroconv.readthedocs.io/en/main/api/nwbconverter.html#neuroconv.nwbconverter.NWBConverter.temporally_align_data_interfaces) to sync my data. However, if it were instead implemented through neuroconv, I could just do [get_event_times_from_ttl()](https://neuroconv.readthedocs.io/en/main/api/interfaces.ecephys.html#neuroconv.datainterfaces.ecephys.spikeglx.spikeglxnidqinterface.SpikeGLXNIDQInterface.get_event_times_from_ttl) and specify the specific subchannel and then sync that to the behavior data using [set_aligned_timestamps()](https://neuroconv.readthedocs.io/en/main/api/basedatainterface.html#neuroconv.basetemporalalignmentinterface.BaseTemporalAlignmentInterface.set_aligned_timestamps).

Here is what the current output is from the neuroconv interface and here is the newer spikeglx feature:

```python
luigi = SpikeGLXConverterPipe(folder_path="./ya008_20240525_g0",)
f = luigi.data_interface_objects['nidq']
print(f.get_channel_names()) #['nidq#XD0']
print(f.get_num_channels()) # 1
```

```python
event = se.read_spikeglx_event("./ya008_20240525_g0")
event.channel_ids # array(['nidq#XD0', 'nidq#XD1', 'nidq#XD2', 'nidq#XD3', 'nidq#XD4', 'nidq#XD5', 'nidq#XD6', 'nidq#XD7'], dtype='<U64')
event.get_num_channels() # 8
```

I envision the interface would be something along the lines of 
```python 
luigi = SpikeGLXConverterPipe(folder_path="./ya008_20240525_g0",)
f = luigi.data_interface_objects['nidq']
trial_start_times = f.recording_extractor.get_traces(channel_ids= f.get_channel_names()[1])  # Gets the first bit that encodes the start times
```


![image](https://github.com/user-attachments/assets/778a75fe-b6a6-4e2b-893e-3f000265ed6b)

**Comment by h-mayorquin** [comment]
Ok, thanks for the explanation. Some questions just to be sure I am following correctly:

First question. 
The current ndiq interface that we have is working with the extractor which returns the multiplexed data. You would like the interface to work with the extractor `read_spikeglx_event` which returns the demultiplexed data. Is that correct?

Second question.
How would you like the data of the nidq to end up in the nwbfile. I am asking this in the context of https://github.com/catalystneuro/neuroconv/issues/994. Reading your last comment it seems that you want the nidq just to synch but I am also wondering if you expect this to end up in the nwbfile in a certain way.

@CodyCBakerPhD, is there any of the synch behavior that you implemented that depends on the multiplexed nature of the data.

Relevant slack conversation (saldly ephemeral):
https://spikesortinggroup.slack.com/archives/C029WTTSP9T/p1723485667692369

**Comment by h-mayorquin** [comment]
> Yes, I would like access to the demultiplexed data. The nidq.meta file we used specifies that the stream is de-mux (niXDChans1=0:7) and so I would like to access it as de-muxed data. Perhaps that is a way to parse the meta that would choose the extractor back end. Open to ideas.

I think that the nidq interface of spikeglx should write demuxed data. I don't understand the second sentence though.  By extractor backend you mean choosing between muxed and demuxed extractor, right? if so, how and why would the meta file have a bearing on that? 

> I think we should include the nidq.bin as raw de-mux since we use it as de-mux data.
OK, I think this should probably be a TimeSeries in acquisition. So mostly like this but not an ElectricalSeries as we are doing right now.

**Comment by tabedzki** [comment]
> I don't understand the second sentence though. By extractor backend you mean choosing between muxed and demuxed extractor, right? if so, how and why would the meta file have a bearing on that?

In the SpikeGLX User manual under the [NIDQ Channels section](https://github.com/billkarsh/SpikeGLX/blob/master/Markdown/UserManual.md#nidq-channels), there are four categories of channels (MN, MA, XA, XD): two (MN, MA)correspond to multiplexed data and two (XA, XD) correspond to non-multiplexed data. By parsing the `nidq.meta` file, one can guess whether the data should be multiplexed and appropriately read the data that way. 

For example, in our experiments, we only set the `XD` channels. Here is a reduced version of our meta file.

```
nSavedChans=1
niMAChans1=
niMNChans1=
niXAChans1=
niXDChans1=0:7
```

Based off this, we know that the only data being sent is of the `XD` type which is non-multiplexed, and it should be read/processed in the non-multiplexed format.

**Comment by h-mayorquin** [comment]
Amazing! Thanks again for the detailed explanation! We can totally do that!

**Comment by CodyCBakerPhD** [comment]
> There is one situation that I do not know: can users (and more importantly do users) mix between multiplexed and non-multiplexed streams for NIDQ?

You would need to mix any time you have at least one analog signal coming in, right? Like running speed on a treadmill, pressure sensors, temperature, etc.?

**Comment by tabedzki** [comment]
I don't know. Can all of those be handled through the NIDQ/SpikeGLX interface? I have no experience using those systems and our lab currently only needs the digital NIDQ pulses for synchronization purposes as far as I am aware. 

If you do believe that they can be mixed between the two, then maybe we should also include the SpikeInterface people about how they process it/allow for both.

**Comment by tabedzki** [comment]
Cody, you are correct. From [Jennifer Colonell](https://spikesortinggroup.slack.com/archives/C029WTTSP9T/p1723740011113269?thread_ts=1723737686.701479&cid=C029WTTSP9T): 

> Yes, you can have any combination.
Multiplexed neural and analog channels require specialized hardware including an external multiplexer system (like the 'Whisper' system developed at Janelia).

**Comment by h-mayorquin** [comment]
Dude, what a joke that those conversations are happening at slack where we will miss them at some point : (

**Comment by tabedzki** [comment]
Should we suggest a forum similar to how the molecular simulation software LAMMPS has a [dedicated forum tracker](https://matsci.org/c/lammps/40)? It is based off [Discourse](https://discourse.org/about). Maybe we could combine the multiple Slack workspaces (Datajoint, SpikeGLX, SpikeInterface, NWB, etc) into a single discourse with different subgroups and have things set up for perpetuity there?

**Comment by CodyCBakerPhD** [comment]
I think Heberto's point is against Slack in general since they are always opt-in, often require an invite, and aren't friendly to web scraping (such as GitHub search)

Slack can be nice to get directly in touch with people or seek private personal help for things you want to keep proprietary or don't otherwise want your questions to be seen by other people on the internet

Discourse just seems to be another public discussion forum (any different from neurostars?) which also doesn't seem much different from GitHub issues/discussions/helpdesks. Point to those is someone (we?) could do quick search attempts to find related conversions to use as references for these kinds of things

**Comment by tabedzki** [comment]
I agree against the ephemeral nature of the Slack and I'll admit to asking questions on Slack which contributes to the problem though as someone relatively new to the community, I didn't even know Neurostars existed. 

The slack workspaces, from what I see, do not mention Neurostars, which would be a good place to be asking these questions. We can try and push the community in that direction, getting buy in from the maintainers/admins of the slack groups to try and push discussions onto something like Neurostars. 

Looking forward to hearing your thoughts on this

**Comment by tabedzki** [comment]
So asking questions for general things like about mixing multiplexing, those would be preferably asked under billkarsh/SpikeGLX/ ? How would encourage other users to ask things not in the Slack but rather on GH?

**Comment by CodyCBakerPhD** [comment]
It's their community so I really have no say in how they want to handle things themselves

But https://github.com/billkarsh/SpikeGLX/ doesn't look like a place for any questions other than specific issues with the main SpikeGLX software

But if I were them I'd make a generic 'SpikeGLX' organization and make a helpdesk repository within it, with discussion enabled. And that would be the go-to place for all general purpose questions and advice (GitHub discussions can even have their own labels/categories to differentiate in-depth technical discussions from general 'how to' questions)

**Comment by tabedzki** [comment]
It's something that can be brought up at the next SpikeSorter monthly meeting, a discussion about how to preserve these discussions and work around free Slack's limitations.

**Comment by CodyCBakerPhD** [comment]
Dang, they work on free slack too? That's especially unfortunate since that 'time out' of message content is especially short until it disappears behind the paywall

**Comment by tabedzki** [comment]
Yes, all the more reason to open up a dialogue on moving to Github discussion or Neurostars.

**Comment by h-mayorquin** [comment]
Unfortunately, it is not a priority right now.

**Comment by yarikoptic** [comment]
@bendichter do you think there could be some workforce to help pushing this issue to a solution?

**Comment by h-mayorquin** [comment]
Hi, @yarikoptic I think this thread got a bit derailed. There are two things to do here:

1) Grant neuroconv access to extract (demuxed) data  from all the modalities described by @tabedzki :

> In the SpikeGLX User manual under the [NIDQ Channels section](https://github.com/billkarsh/SpikeGLX/blob/master/Markdown/UserManual.md#nidq-channels), there are four categories of channels (MN, MA, XA, XD): two (MN, MA)correspond to multiplexed data and two (XA, XD) correspond to non-multiplexed data.

2) Write that data into NWB. 

We don't really have resources to attack this problem as a whole at the moment but if you have some specific category that you want to access through neuroconv we might be able to allocate some time to start working on that. If so, I have two questions:

1) Which channel category.
2) How do want the data to be written? Do you guys want the raw data to be written or expect this to be written as an Events object. The first is easier, the second one probably would be slower as we are waiting on some release of ndx-events at the moment.

**Comment by manimoh** [comment]
Hi @h-mayorquin 
1. As things stand, the data from XA channels is being consumed properly by `SpikeGLXNIDQInterface`, but not from the XD channels. So we need the data from XD channels to be extracted and fed into the NWB file.
2. Writing the data as events object would be ideal.

**Comment by h-mayorquin** [comment]
Hi @manimoh,

Yeah, maybe we can add support the XD channels.

Do you have some testing data that I can use to build against? You probably can find me on the dandi slack if you want to discuss this in more detail.

**Comment by manimoh** [comment]
Hi @h-mayorquin, 
I had uploaded this [dataset ](https://gin.g-node.org/NeuralEnsemble/ephy_testing_data/src/master/spikeglx/DigitalChannelTest_g0) to build the neo-support for digital events in SpikeGLX, so maybe you can use that?

And could you send me an invite for Dandi slack to manishofyore@gmail.com

**Comment by yarikoptic** [comment]
@h-mayorquin @bendichter ping on this issue -- is there anything we could do on our end to get to the next steps here?

**Comment by h-mayorquin** [comment]
Hey guys, looking into the data that you shared, is it mean to have data from only one digital channel?

```python
from spikeinterface.extractors import SpikeGLXRecordingExtractor, SpikeGLXEventExtractor
from pathlib import Path

folder_path = Path("/home/heberto/neuroconv_testing_data/ephy_testing_data/spikeglx/DigitalChannelTest_g0")
assert folder_path.exists()

recording = SpikeGLXEventExtractor(folder_path=folder_path)
for channel_id in recording.channel_ids:
    event_times = recording.get_event_times(channel_id=channel_id)
    print(f"{channel_id=}, {len(event_times)=}")


channel_id='nidq#XD0', len(event_times)=326
channel_id='nidq#XD1', len(event_times)=0
channel_id='nidq#XD2', len(event_times)=0
channel_id='nidq#XD3', len(event_times)=0
channel_id='nidq#XD4', len(event_times)=0
channel_id='nidq#XD5', len(event_times)=0
channel_id='nidq#XD6', len(event_times)=0
channel_id='nidq#XD7', len(event_times)=0
```
Is this correct and what you expected?

**Comment by manimoh** [comment]
Yes this is correct, and what I would expect.
